File: javascript/dsa/algorithms/strings/group-anagrams.js
function groupAnagrams(strs) {
  const map = new Map();
  
  for (const s of strs) {
    const freq = new Array(26).fill(0);
    for (const c of s) {
      freq[c.charCodeAt(0) - 97]++;
    }
    
    let key = '';
    for (let i = 0; i < 26; i++) {
      if (freq[i] > 0) {
        key += String.fromCharCode(i + 97) + freq[i];
      }
    }
    
    if (!map.has(key)) {
      map.set(key, []);
    }
    map.get(key).push(s);
  }
  
  return Array.from(map.values());
}

File: cpp/dsa/algorithms/strings/group-anagrams.cpp
vector<vector<string>> groupAnagrams(vector<string>& strs) {
  unordered_map<string, vector<string>> mp;
  
  for (string s : strs) {
    vector<int> freq(26, 0);
    for (char c : s) {
      freq[c - 'a']++;
    }
    
    string key = "";
    for (int i = 0; i < 26; i++) {
      if (freq[i] > 0) {
        key += (char)('a' + i) + to_string(freq[i]);
      }
    }
    
    if (mp.find(key) == mp.end()) {
      mp[key] = {};
    }
    mp[key].push_back(s);
  }
  
  return resultFromMap(mp);
}